dss/stepfunctions/visitation/__init__.py

```diff
@@ -0,0 +1,161 @@
+    walker_state_spec: dict = dict()
+
+    def __init__(self, state_spec: dict, state: dict, logger) -> None:
+        """
+        Pull in fields defined in state specifications and set as instance properties
+        """
+        self.state_spec = state_spec
+        state = copy.deepcopy(state)
+
+        for k, default in state_spec.items():
+            v = state.get(k, None)
+
+            if v is None:
+                if callable(default):
+                    v = default()
+                else:
+                    v = default
+
+            setattr(self, k, v)
+
+        self.logger = logger
+
+    @classmethod
+    def _with_state(cls, state: dict, logger) -> 'Visitation':
+        """
+        Pull in state specific to the job.
+        """
+        state_spec = {
+            ** Visitation._state_spec,
+            ** cls.state_spec,
+            ** cls.walker_state_spec,
+        }
+        return cls(state_spec, state, logger)
+
+    def get_state(self) -> dict:
+        """
+        Return step function state at the end of each lambda defined in the job and walker step functions.
+        """
+        return {
+            k: getattr(self, k)
+            for k in self.state_spec
+        }
+
+    @classmethod
+    def start(cls, replica: str, bucket: str, number_of_workers: int) -> str:
+        name = '{}--{}'.format(cls.__name__, str(uuid4()))
+
+        inp = {
+            '_visitation_class_name': cls.__name__,
+            'replica': replica,
+            'bucket': bucket,
+            '_number_of_workers': number_of_workers,
+        }
+
+        step_functions_invoke('dss-visitation-{stage}', name, inp)
+
+        return name
+
+    def job_initialize(self) -> None:
+        """
+        Implement for initialization or sanity checking for a job.
+        """
+        pass
+
+    def job_finalize(self) -> None:
+        """
+        Implement for finalization work for a successful job. Called once each worker has completed.
+        """
+        pass
+
+    def job_finalize_failed(self) -> None:
+        """
+        Implement for finalization work for a failed job. This is your opportunity to cry, notify, and ruminate.
+        """
+        pass
+
+    def walker_initialize(self) -> None:
+        """
+        Implement this method for initialization or sanity checking specifically for a walker.
+        """
+        pass
+
+    def walker_walk(self) -> None:
+        """
+        Subclasses must implement this method. Called for walker thread.
+        """
+        raise NotImplementedError
+
+    def walker_finalize(self) -> None:
+        """
+        Implement this method for finalization work specific to a walker.
+        """
+        pass
+
+    def walker_finalize_failed(self) -> None:
+        """
+        Aliment this method for finalization work specific to a failed walker.
+        """
+        pass
+
+    def __setattr__(self, key, val):  # typing.type: (str, typing.Any) -> None
```

Comment:
Please remove these. LMK if you need help sorting out mypy behavior.

Reply:
Removed base class, but leaving __setattr__ and __getattr__ for now, with a link to MyPy recommendations regarded mixed in properties.

Reply:
👍